File: go/maybe.go
package main

import (
	"fmt"
)


type a interface {}
type r interface {}

type Maybe struct {
    maybe   func(r, func(a) r) r
}

var nothing Maybe = Maybe{func(n r, j func(a) r) r {return n}}

func just(x a) Maybe {
    return Maybe{func(n r, j func(a) r) r {return j(x)}}
}

func (maybe Maybe) String() string {
    return maybe.maybe("Nothing", func(x a) r {return fmt.Sprintf("Just %v", x)}).(string)
}

func (maybe Maybe) Map(func(a) b) 

func main() {	
	var x = just(1)
	fmt.Println(x)
}

File: scala/src/Maybe.scala
//> using scala 3.3.1

package boehmberarducci

trait Maybe[+A]:
  def fold[R](nothing: => R)(just: A => R): R

  def map[B](f: A => B): Maybe[B] =
    fold[Maybe[B]](Nothing)(x => Just(f(x)))

  def flatMap[B](f: A => Maybe[B]): Maybe[B] =
    fold[Maybe[B]](Nothing)(f)

  def getOrElse[B >: A](default: => B): B =
    fold[B](default)(identity)

case object Nothing extends Maybe[Nothing]:
  def fold[R](nothing: => R)(just: Nothing => R): R = nothing

case class Just[A](value: A) extends Maybe[A]:
  def fold[R](nothing: => R)(just: A => R): R = just(value)
